runtime/src/accounts_db.rs

```diff
@@ -3351,16 +3387,44 @@ impl AccountsDb {
     fn write_accounts_to_cache(
         &self,
         slot: Slot,
-        hashes: &[Hash],
+        hashes: Option<&[Hash]>,
         accounts_and_meta_to_store: &[(StoredMeta, &AccountSharedData)],
     ) -> Vec<AccountInfo> {
-        assert_eq!(hashes.len(), accounts_and_meta_to_store.len());
-        accounts_and_meta_to_store
-            .iter()
-            .zip(hashes)
-            .map(|((meta, account), hash)| {
-                self.accounts_cache
-                    .store(slot, &meta.pubkey, (**account).clone(), *hash);
+        let len = accounts_and_meta_to_store.len();
+        let empty = vec![];
+        let hashes = match hashes {
+            Some(hashes) => {
+                assert_eq!(hashes.len(), len);
+                hashes
+            }
+            None => &empty,
+        };
+
+        (0..len)
+            .into_iter()
+            .map(|i| {
+                let (meta, account) = &accounts_and_meta_to_store[i];
```

Comment:
In general manual indexing is slightly slower in rust because the compiler cannot remove the bounds checks. Why we are using the iterator here originally.
Can use enumerate for the index for `hashes`